packages/api-client/src/connections/ApiConnection.ts

```diff
@@ -0,0 +1,108 @@
+/*
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ *
+ */
+import nodeFetch from 'node-fetch';
+import { QueryParameters, AuthHandler } from '../types';
+import type { RequestInit, HeadersInit, Response } from 'node-fetch';
+import { stringify } from 'qs';
+
+type RequestBody = Record<string, unknown> | Record<string, unknown>[];
+
+type FetchHeaders = HeadersInit & {
+  Authorization: string;
+  'Content-Type'?: string;
+}
+
+type FetchConfig = RequestInit & {
+  headers: FetchHeaders;
+}
+
+const DEFAULT_MAX_WAIT_TIME = 45 * 1000; // 45 seconds in milliseconds
+
+/**
+ * TODO: make internal, do not export from package
+ *
+ *
+ * TODO: restore other ApiConnection, let other packages use that. This one is different.
+ */
+export class ApiConnection {
+  private readonly authHandler: AuthHandler;
+
+  private readonly baseUrl: string;
+
+  constructor(authHandler: AuthHandler, baseUrl: string) {
+    this.authHandler = authHandler;
+    this.baseUrl = baseUrl;
+  }
+
+  async get(endpoint: string, queryParameters?: QueryParameters | null) {
+    return this.request('GET', endpoint, queryParameters);
+  }
+
+  async post(endpoint: string, queryParameters?: QueryParameters | null, body?: RequestBody | null) {
+    return this.request('POST', endpoint, queryParameters, body);
+  }
+
+  async put(endpoint: string, queryParameters?: QueryParameters | null, body?: RequestBody | null) {
+    return this.request('PUT', endpoint, queryParameters, body);
+  }
+
+  async delete(endpoint: string) {
+    return this.request('DELETE', endpoint);
+  }
+
+  private async request(
+    requestMethod: string,
+    endpoint: string,
+    queryParameters: QueryParameters = {},
+    body?: RequestBody,
+  ) {
+    const queryUrl = this.stringifyQuery(this.baseUrl, endpoint, queryParameters);
+    const fetchConfig: FetchConfig = {
+      method: requestMethod || 'GET',
+      headers: {
+        Authorization: this.authHandler.getAuthHeader(),
+        'Content-Type': 'application/json',
+      },
+    };
+    if (body) {
+      fetchConfig.body = JSON.stringify(body);
+    }
+
+    const response = await this.fetchWithTimeout(queryUrl, fetchConfig);
+    await this.verifyResponse(response);
+    return response.json();
+  }
+
+  private async fetchWithTimeout(url: string, config: RequestInit, timeout: number = DEFAULT_MAX_WAIT_TIME): Promise<Response> {
+    return nodeFetch(url, {...config, timeout } );
```

Comment:
Reimplemented this method to avoid importing from utils. Because we are publishing this package, we can't have it depend on @tupaia/* otherwise the whole of tupaia will be shipped with the package.